library-best-practices/JLBP-6.md

```diff
@@ -0,0 +1,152 @@
+[JLBP-6] Avoid creating conflicts between major versions
+------------------------------------------------------------------------
+
+- A new major version should generally use a different Java package.
+- There is a tradeoff: if an extremely large library (for example thousands of
+  classes) breaks a single method, the cost of a package rename may be higher
+  than dealing with the fallout of dealing with diamond dependency conflicts for
+  that particular method. Only perform an in-place breakage if users can easily
+  replace their usage of the old library with the new library. The larger the
+  portion of a library that breaks between major versions, the higher the value
+  is from renaming the java package.
+- *If* the Java package is changed, use a different Maven ID (different group ID
+  or artifact ID)
+- *If* the Java package is kept the same, use the same Maven ID (same group ID
+  and same artifact ID).
+- In other words, rename the Java package if and only if the Maven ID changes.
+
+Background
+----------
+
+When a library B in Java depends on another library A through the Maven
+repository system, there are two handles that library B needs to use to access
+library A:
+
+1. The Maven coordinates, following the form
+   `${groupId}:${artifactId}:${version}`, for example
+   `com.google.guava:guava:26.0-jre`. The Maven coordinates are used to pull the
+   library's jar from Maven Central. Given a group ID and artifact ID, Maven
+   will choose only one version to pull in.
+2. The Java package (the package used in `import` statements), for example
+   `com.google.common.collect.ImmutableList`. The Java package is used to select
+   a particular class within the jar pulled from Maven Central.
+
+Now when breaking changes are introduced to library A between major version 1
+and major version 2, a choice needs to be made: to rename or not rename? This
+question applies to both items listed above, the Maven ID (1) and the Java
+package (2). Consider the following renaming scenario:
+
+- The Maven coordinates start as `g1:a1:1.0.0`
+  - With no rename of the Maven ID, the new major version's Maven coordinates
+    are `g1:a1:2.0.0`. (Only one of the versions can be present in the
+    dependency tree.)
+  - With a rename of the Maven ID, the new major version's Maven coordinates are
+    `g2:a2:2.0.0`. (Both versions can be present in the dependency tree.)
+- The Java package starts as `com.google.a1`
+  - With no rename, the new major version's Java package is still
+    `com.google.a1`
+  - With a rename, the new major version's Java package is `com.google.a2`
+
+Given this scenario, here are the possible combinations of renamings:
+
+- **Don't rename Java package**:
+  - **Don't rename Maven ID (case 1)**: This approach can result in diamond
```

Comment:
Honestly, this can happen even in minor version updates too; for instance if your library depends on a new, non-breaking method addition introduced in a new minor version.